File: 3.Stacks and Queues/1.implementation_queue.js
class Node {
    constructor(value) {
      this.value = value;
      this.next = null;
    }
  }
  
  class QueueLinkedList {
    constructor(){
      this.first = null;
      this.last = null;
      this.length = 0;
    }
    peek() {
        return this.first;
    }
    enqueue(value){
      const node = new Node(value);
      if(this.length === 0){
        this.first = node;
        this.last = node;
      } else {
        this.last.next = node;
        this.last = node;
      }
      this.length++;
      return this;
    }
    dequeue(){
        if(!this.first){ // or this.length === 0
            return null;
        }
        if(this.first === this.last){ // or this.length === 1
            this.last = null;
        }
        const temp = this.first;
        this.first = this.first.next;
        this.length--;
        return temp;
    }
    isEmpty(){
      return this.length === 0;
    }
  }
  
  const myQueue = new QueueLinkedList();
  myQueue.enqueue('Joy');
  myQueue.enqueue('Matt');
  myQueue.enqueue('Pavel');
  myQueue.enqueue('Samir');
  console.log(myQueue.peek());
  console.log(myQueue.dequeue());
  console.log(myQueue.dequeue());
  console.log(myQueue.peek());
  
  

File: 3.Stacks and Queues/1.implementation_queue.py
class Node:
    def __init__(self, value: int) -> None:
        self.value = value
        self.next = None


class QueueLinkedList:
    def __init__(self) -> None:
        self.first = None
        self.last = None
        self.length = 0

    def peek(self):
        return self.first.value
    
    def enqueue(self, value):
        node = Node(value)
        if self.length == 0:
            self.first = node
            self.last = node
        else:
            self.last.next = node
            self.last = node
        
        self.length += 1
        return self
    
    def dequeue(self):
        if self.length == 0:
            return self
        if self.length == 1:
            self.first = None
            self.last = None
        else:
            tmp = self.first
            self.first = self.first.next
            
        self.length -= 1
        return tmp
    